### ISSUE TITLE
The locator is present but using the Winappdriver the bwlow error is thrown

### ISSUE THREAD

**Issue by deepabhatk** [microsoft/WinAppDriver#1805]

**Comment by deepabhatk** [comment]
I am trying to get the text of specific element which has automation id as "RecordingIconButton Timerlabel".It is the child element and we are trying to use the xpath for locating it. But element not found  Webdriver exception is  thrown for the same . Can anyone suggest any solution for the same


Locator used -Xpath -"//button[contains(@AutomationId,'RecordingButton')]/*[1]"

![Doubht](https://user-images.githubusercontent.com/33826746/195290987-6876b8f9-3a47-43b9-8aef-38e628e8c7da.PNG)

**Comment by anunay1** [comment]
What is the exception? Have you tried using any other locator other than xpath?

**Comment by deepabhatk** [comment]
> What is the exception? Have you tried using any other locator other than xpath? There is a know issue with xpath for being slow.

Yes i tried with the AutomationId  as well- 'RecordingIconButtonTimerLabel' .Still the issue persist.

I am getting the following Exception:

   OpenQA.Selenium.WebDriverException: An element could not be located on the page using the given search parameters.
    TestCleanup method Automation.TestApp.SpecFlow.Features.Screen_Features.AutobalanceFeature.TestTearDown threw exception. OpenQA.Selenium.WebDriverException: OpenQA.Selenium.WebDriverException: An element could not be located on the page using the given search parameters..



 Stack Trace:
    RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
    RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
    AppiumDriver`1.Execute(String driverCommandToExecute, Dictionary`2 parameters)
    RemoteWebDriver.FindElement(String mechanism, String value)
    AppiumDriver`1.FindElement(String by, String value)
    AppiumDriver`1.FindElementByAccessibilityId(String selector)
    Element_Interactions.FindElement(String locator) line 57
    Element_Interactions.GetAttribute(String element, String attribute) line 531
    Surgical_View_Header_Interactions.GetNameAttributeOfTimer() line 180
    Surgical_View_Header_Steps.VerifyTimerEnabled(String status) line 362
    lambda_method(Closure , IContextManager , String )
    BindingInvoker.InvokeBinding(IBinding binding, IContextManager contextManager, Object[] arguments, ITestTracer testTracer, TimeSpan& duration) line 69
    TestExecutionEngine.ExecuteStepMatch(BindingMatch match, Object[] arguments) line 521
    TestExecutionEngine.ExecuteStep(IContextManager contextManager, StepInstance stepInstance) line 426
    TestExecutionEngine.OnAfterLastStep() line 207
    TestRunner.CollectScenarioErrors() line 59
    AutobalanceFeature.ScenarioCleanup()
    AutobalanceFeature._06VerifyThatTheUserIsAbleToStartStopTheVideoRecordingsFromGUI_() line 54
    TestCleanup Stack Trace
    RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
    RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
    AppiumDriver`1.Execute(String driverCommandToExecute, Dictionary`2 parameters)
    RemoteWebDriver.FindElement(String mechanism, String value)
    AppiumDriver`1.FindElement(String by, String value)
    AppiumDriver`1.FindElementByAccessibilityId(String selector)

**Comment by anunay1** [comment]
Can you share what is shown in the win app driver console?

**Comment by ChristoWolf** [comment]
As you can see in your inspect.exe output, the element's `IsControlElement` property is false, i.e. it is not exposed to the control automation tree.
WinAppDriver only interacts with that subtree though, so you need to expose that element accordingly.
You can achieve that by overriding the element's `AutomationPeer`'s [`IsControlElementCore`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.automation.peers.automationpeer.iscontrolelementcore?view=windowsdesktop-6.0) method so that it always returns true.

**Comment by deepabhatk** [comment]
Hi @ChristoWolf 
Thank you for your input. Can you please let me know or give some reference of setting the same. I can see that it comes from presentation.dll. So the changes are to made in design or in Winappdriver we have some way to access such elements.

**Comment by ChristoWolf** [comment]
Sure!

1. [Customize the element's `AutomationPeer` by deriving from it](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/ui-automation-of-a-wpf-custom-control?view=netframeworkdesktop-4.8#customizations-in-a-derived-peer).
2. [Override `IsControlElementCore` in your custom peer such that 9it always returns `true`](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/ui-automation-of-a-wpf-custom-control?view=netframeworkdesktop-4.8#override-core-methods).
3. There is even exactly this use case mentioned there in one of the paragraphs.
4. [Customize the element by overriding its `OnCreateAutomationPeer` method such that it returns your custom `AutomationPeer`](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/ui-automation-of-a-wpf-custom-control?view=netframeworkdesktop-4.8#override-oncreateautomationpeer).
5. In your XAML file, use your fully customized element instead of the base one which was not exposed to the automation tree.
